File: app/src/main/java/org/stepic/droid/ui/activities/PhotoViewActivity.java
package org.stepic.droid.ui.activities;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.MenuItem;

import org.jetbrains.annotations.Nullable;
import org.stepic.droid.R;
import org.stepic.droid.base.SingleFragmentActivity;
import org.stepic.droid.ui.fragments.PhotoViewFragment;

public class PhotoViewActivity extends SingleFragmentActivity {

    public static final String pathKey = "pathKey";
//    private GestureDetector gestureDetector;

    @Nullable
    @Override
    protected Fragment createFragment() {
        String path = getIntent().getStringExtra(pathKey);
        return PhotoViewFragment.newInstance(path);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.no_transition, R.anim.slide_out_to_bottom);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        gestureDetector = new GestureDetector(this, new SwipeDetector());
    }

//    private class SwipeDetector extends GestureDetector.SimpleOnGestureListener {
//        private static final int SWIPE_MIN_DISTANCE = 120;
//        private static final int SWIPE_MAX_OFF_PATH = 250;
//        private static final int SWIPE_THRESHOLD_VELOCITY = 200;
//
//        @Override
//        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
//
//            if (Math.abs(e1.getX() - e2.getX()) > SWIPE_MAX_OFF_PATH)
//                return false;
//
//            if (e2.getY() - e1.getY() > SWIPE_MIN_DISTANCE && Math.abs(velocityY) > SWIPE_THRESHOLD_VELOCITY) {
//                finish();
//                return true;
//            }
//
//            if (e1.getY() - e2.getY() > SWIPE_MIN_DISTANCE && Math.abs(velocityY) > SWIPE_THRESHOLD_VELOCITY) {
//                finish();
//                overridePendingTransition(R.anim.no_transition, R.anim.slide_out_to_top);
//                return true;
//            }
//
//            return false;
//        }
//    }
//
//    @Override
//    public boolean dispatchTouchEvent(MotionEvent ev) {
//        // TouchEvent dispatcher.
//        if (gestureDetector != null) {
//            if (gestureDetector.onTouchEvent(ev))
//                // If the gestureDetector handles the event, a swipe has been
//                // executed and no more needs to be done.
//                return true;
//        }
//        return super.dispatchTouchEvent(ev);
//    }
//
//    @Override
//    public boolean onTouchEvent(MotionEvent event) {
//        return gestureDetector.onTouchEvent(event);
//    }
}

File: app/src/main/java/org/stepic/droid/ui/activities/PhotoViewActivity.kt
package org.stepic.droid.ui.activities

import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.graphics.drawable.PictureDrawable
import android.os.Bundle
import android.view.MenuItem
import androidx.core.view.isVisible
import dev.androidbroadcast.vbpd.viewBinding
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.github.chrisbanes.photoview.PhotoViewAttacher
import org.stepic.droid.R
import org.stepic.droid.base.FragmentActivityBase
import org.stepic.droid.databinding.FragmentPhotoViewBinding
import org.stepik.android.view.glide.model.GlideRequestFactory
import kotlin.math.abs
import kotlin.math.min

class PhotoViewActivity : FragmentActivityBase() {
    companion object {
        const val EXTRA_PATH = "extra_path"
    }

    private val photoViewBinding: FragmentPhotoViewBinding by viewBinding(FragmentPhotoViewBinding::bind)
    private lateinit var photoViewAttacher: PhotoViewAttacher
    private var screenHeight: Int = 0
    private var dismissPathLength: Int = 0

    private val target = object : CustomTarget<PictureDrawable>() {
        override fun onResourceReady(resource: PictureDrawable, transition: Transition<in PictureDrawable>?) {
            photoViewBinding.internetProblemRootView.isVisible = false
            photoViewBinding.zoomableImageView.setImageDrawable(resource)
            photoViewAttacher.update()
//            Timber.d("resource ready")
        }

        override fun onLoadCleared(placeholder: Drawable?) {

        }

        override fun onLoadFailed(errorDrawable: Drawable?) {
            photoViewBinding.internetProblemRootView.isVisible = true
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_photo_view)
        setUpToolbar()
        photoViewAttacher = PhotoViewAttacher(photoViewBinding.zoomableImageView)
        screenHeight = Resources.getSystem().displayMetrics.heightPixels
        dismissPathLength = resources.getDimensionPixelSize(R.dimen.dismiss_path_length)

        photoViewBinding.verticalDragLayout.setOnDragListener { dy ->
            if (photoViewAttacher.scale > 1f) return@setOnDragListener
            val alpha = 1 - min(abs(dy / (3 * dismissPathLength)), 1f)
            photoViewBinding.backgroundColorView.alpha = alpha
            photoViewBinding.toolbar.alpha = alpha
            photoViewBinding.zoomableImageView.translationY = -dy
        }

        photoViewBinding.verticalDragLayout.setOnReleaseDragListener { dy ->
            if (photoViewAttacher.scale > 1f) return@setOnReleaseDragListener
            if (abs(dy) > dismissPathLength) {
                photoViewBinding.zoomableImageView.isVisible = false
                finish()
            } else {
                photoViewBinding.backgroundColorView.alpha = 1f
                photoViewBinding.toolbar.alpha = 1f
                photoViewBinding.zoomableImageView.translationY = 0f
            }
        }

        val url = requireNotNull(intent.getStringExtra(EXTRA_PATH))

        photoViewBinding.retryButton.setOnClickListener {
            photoViewBinding.internetProblemRootView.isVisible = false
            loadImage(url)
        }
        loadImage(url)
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.no_transition, R.anim.no_transition)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }
            else ->
                super.onOptionsItemSelected(item)
        }

    private fun loadImage(url: String) {
        GlideRequestFactory
            .create(this, null)
            .load(url)
            .fitCenter()
            .into(target)
    }

    private fun setUpToolbar() {
        setSupportActionBar(photoViewBinding.toolbar)
        val supportActionBar = supportActionBar

        if (supportActionBar != null) {
            supportActionBar.setDisplayShowTitleEnabled(false)
            supportActionBar.setDisplayShowHomeEnabled(true)
            supportActionBar.setHomeButtonEnabled(true)
            supportActionBar.setDisplayHomeAsUpEnabled(true)
            supportActionBar.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)
        }
    }
}
